src/konserve/cache.cljc

```diff
@@ -21,84 +23,131 @@
    is not present, or the not-found value if supplied."
   ([store key]
    (get store key nil))
   ([store key not-found]
-   (get-in store [key] not-found)))
+   (get store key not-found {:sync? false}))
+  ([store key not-found opts]
+   (trace "get on key " key opts)
+   (get-in store [key] not-found opts)))
 
 (defn update-in
   "Updates a position described by key-vec by applying up-fn and storing
   the result atomically. Returns a vector [old new] of the previous
   value and the result of applying up-fn (the newly stored value)."
-  [store key-vec up-fn & args]
-  (go-locked
-   store (first key-vec)
-   (let [cache (:cache store)
-         [old new] (<! (-update-in store key-vec (partial meta-update (first key-vec) :edn) up-fn args))]
-     (swap! cache cache/evict (first key-vec))
-     [old new])))
+  ([store key-vec up-fn]
+   (update-in store key-vec up-fn {:sync? false}))
+  ([store key-vec up-fn opts]
+   (trace "update-in on key " key opts)
+   (async+sync (:sync? opts)
+               *default-sync-translation*
+               (go-locked
+                store (first key-vec)
+                (let [cache (:cache store)
+                      key (first key-vec)
+                      [old new] (<! (-update-in store key-vec (partial meta-update (first key-vec) :edn) up-fn opts))]
+                  (when (cache/has? @cache key)
+                    (swap! cache cache/evict key)
+                    (swap! cache cache/miss key new))
+                  [old new])))))
 
 (defn update
   "Updates a position described by key by applying up-fn and storing
   the result atomically. Returns a vector [old new] of the previous
   value and the result of applying up-fn (the newly stored value)."
-  [store key meta-up-fn fn & args]
-  (apply update-in store [key] meta-up-fn fn args))
+  ([store key fn]
+   (update store key fn {:sync? false}))
+  ([store key fn opts]
+   (trace "update on key " key opts)
+   (update-in store [key] fn opts)))
 
 (defn assoc-in
   "Associates the key-vec to the value, any missing collections for
   the key-vec (nested maps and vectors) are newly created."
-  [store key-vec val]
-  (go-locked
-   store (first key-vec)
-   (let [cache (:cache store)]
-     (swap! cache cache/evict (first key-vec))
-     (<! (-assoc-in store key-vec (partial meta-update (first key-vec) :edn) val)))))
+  ([store key-vec val]
+   (assoc-in store key-vec val {:sync? false}))
+  ([store key-vec val opts]
+   (trace "assoc-in on key " key opts)
+   (async+sync (:sync? opts)
+               *default-sync-translation*
+               (go-locked
+                store (first key-vec)
+                (let [cache (:cache store)
+                      [old new] (<! (-assoc-in store key-vec (partial meta-update (first key-vec) :edn) val opts))]
+                  (swap! cache cache/evict (first key-vec))
+                  (swap! cache cache/miss (first key-vec) new)
+                  [old new])))))
```

Comment:
Missing `(when (cache/has? @cache key)` like in `update-in`

Reply:
Done.